### ISSUE TITLE
Unsupported long video upload (>30s) because of lack media_category param

### ISSUE THREAD

**Issue by Dyukon** [ttezel/twit#306]
Hi,

At the moment Twitter allows uploading of long videos (up to 140 seconds length) using async media upload mechanism. For that it's mandatory to specify parameter media_category='tweet_video' (look here: https://twittercommunity.com/t/media-upload-api-does-not-allow-video-more-than-30-seconds/69665).
Unfortunately, media_category parameter isn't supported by FileUploader now, so long videos uploading is impossible.

**Comment by planetamongoloid** [comment]
I have the same problem! is there any workaround for this?

**Comment by Dyukon** [comment]
At the moment I made all files from 'twit' module a part of my project and edited one of them manually in two places:

_file_uploader.js_
================
var FileUploader = function (params, twit) {
  assert(params)
  assert(params.file_path, 'Must specify `file_path` to upload a file. Got: ' + params.file_path + '.')
  var self = this;
  self._file_path = params.file_path;
  self._twit = twit;
  self._isUploading = false;
  self._isFileStreamEnded = false;
  **self._media_category = params.media_category;**
}
...
FileUploader.prototype._initMedia = function (cb) {
  var self = this;
  var mediaType = mime.lookup(self._file_path);
  var mediaFileSizeBytes = fs.statSync(self._file_path).size;

  // Check the file size - it should not go over 15MB for video.
  // See https://dev.twitter.com/rest/reference/post/media/upload-chunked
  if (mediaFileSizeBytes < MAX_FILE_SIZE_BYTES) {
    **var params = {
      'command': 'INIT',
      'media_type': mediaType,
      'total_bytes': mediaFileSizeBytes
    };
    if (self._media_category) {
        params.media_category = self._media_category;
    }
    debug('params: ' + JSON.stringify(params));
    self._twit.post('media/upload', params, cb);**
  } else {
    var errMsg = util.format('This file is too large. Max size is %dB. Got: %dB.', MAX_FILE_SIZE_BYTES, mediaFileSizeBytes);
    cb(new Error(errMsg));
  }
}

**Comment by CTassisF** [comment]
@Dyukon That isn't enought.
When using media_category you have to check for the status returned with [FINALIZE](https://dev.twitter.com/rest/reference/post/media/upload-finalize) and eventually call [GET media/upload (STATUS)](https://dev.twitter.com/rest/reference/get/media/upload-status) many times until de processing_info.state is either success or failed. If you try to use the media_id in a tweet before its status is "success" you will get "Not valid video" error.

**Comment by CTassisF** [comment]
@Dyukon Yesterday I tried to implement this in [my fork](https://github.com/CTassisF/twit). It seems to work partially, the only problem I got was that [this](https://github.com/CTassisF/twit/blob/master/lib/file_uploader.js#L89-L105) was always returning statusCode 4xx and empty "status". I don't know what I did wrong.

**Comment by Dyukon** [comment]
@CTassisF 
You're absolutely right. To place the video into tweet we need to wait until it's processed on the tweet server. I forgot to include this part of code here. I do it following way (in my app, outside changed file_uploader):

```
                if (mediaId) {
                    // wait until uploaded video is processed
                    let processingInfo = res.processing_info;
                    while (processingInfo.state==='pending') {
                        yield Promise.delay(processingInfo.check_after_secs * 1000);
                        let res = yield req.twitter.get('/media/upload', {
                            command: 'STATUS',
                            media_id: mediaId
                        });
                        processingInfo = res.data.processing_info;
                    }
                }
```

**Comment by planetamongoloid** [comment]
thanks! I'll try that.

**Comment by xerosanyam** [comment]
any update on this ?

**Comment by koji98** [comment]
Hey, I'm confused with the second part, I don't use res anywhere in my bot so I am not sure what to do
